server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java

```diff
@@ -160,9 +162,11 @@ public static List<DfsKnnResults> mergeKnnResults(SearchRequest request, List<Df
             }
         }
 
-        List<DfsKnnResults> mergedResults = new ArrayList<>(request.source().knnSearch().size());
-        for (int i = 0; i < request.source().knnSearch().size(); i++) {
-            TopDocs mergedTopDocs = TopDocs.merge(request.source().knnSearch().get(i).k(), topDocsLists.get(i).toArray(new TopDocs[0]));
+        List<DfsKnnResults> mergedResults = new ArrayList<>(source.knnSearch().size());
+        for (int i = 0; i < source.knnSearch().size(); i++) {
+            Integer topK = source.knnSearch().get(i).k();
+            int size = source.size() == -1 ? DEFAULT_SIZE : source.size();
+            TopDocs mergedTopDocs = TopDocs.merge(topK != null ? topK : size, topDocsLists.get(i).toArray(new TopDocs[0]));
```

Comment:
I wonder if we could handle this within the SearchSourceBuilder directly? When parsing (thus creating via an API), parsing the `knn` section or `size` section could set the `k` for the `knn` there.

Similarly if somebody calls `SearchSourceBuilder#setSize(int)` that could propagate down to the `knn` object if it exists?

Reply:
I was looking into it initially, but I wasn't impressed with a couple of approaches I've tried, as we were introducing dependencies between different component in the `SearchSourceBuilder` - especially when there's no guaranty of order.  On the other hand, as things are now, if someone calls `SearchSourceBuilder#setSize` at a later point, then they would also have the responsibility of setting up & updating `k` which is not right 🤔  (similarly for `org.elasticsearch.search.DefaultSearchContext#size(int) `) 

Will look into it a bit more.

Reply:
Updated the functionality to go through the `SearchSourceBuilder` in https://github.com/elastic/elasticsearch/pull/101209/commits/9092f378c9c8ad8d2b5aebc661bf2232fd712160. Haven't set it in the `SearchSourceBuilder#size` method, but rather when we parse the initial `size` param (iff present).  

Happy to update it as well, but was wondering which cases related to `knn` search we'd capture (was looking at the usages but I'm not very familiar with all cases).